pipeline/photo_cal.py

```diff
@@ -28,12 +70,185 @@ class PhotCalibrator:
+            else:
+                # If there are more than one, just pick the first one in the list
+                # that matches.
+                # TODO : think about whether we can do better.
+                catdex.append( i )
+                sourcedex.append( w[0] )
+
+        _logger.debug( f"Matched {len(catdex)} stars between Gaia and the image source list" )
+
+        if len(catdex) < min_matches:
+            raise BadMatchException( f"Only matched {lencatdex} stars between Gaia and the image source list, "
+                                     f"which is less than the minimum of {min_matches}" )
+
+        sourcera = sourcera[ sourcedex ]
+        sourcedec = sourcedec[ sourcedex ]
+        sourceflux = sourceflux[ sourcedex ]
+        sourcefluxerr = sourcefluxerr[ sourcedex ]
+        catdata = catdata[ catdex ]
+
+        # At this point we have an Astropy Table in catdata with the
+        # catalog information, and we have the image source information
+        # arrays index-matched to catdata (sourcera, sourcedec,
+        # sourceflux, sourcefluxerr)
+
+        # Pull the GaiaDR3 MAG_G to instrument filter transformation
+        # from the Instrument object.
+
+        trns = image.instrument_object.get_GaiaDR3_transformation( image.filter_short )
+
+        # mag = -2.5*log10(flux) + zp
+        # mag = GaiaGmag - sum( trns[i] * ( GaiaBP - GaiaRP )**i )
+        # ...so...
+        # zp = GaiaGmag - sum( trns[i] * ( GaiaBP - GaiaRP )**i ) + 2.5*log10( flux )
+
+        fitorder = len(trns) - 1
+
+        cols = catdata[ 'MAG_BP' ] - catdata[ 'MAG_RP' ]
+        colerrs = np.sqrt( catdata[ 'MAGERR_BP' ]**2 + catdata[ 'MAGERR_RP' ]**2 )
+        colton = cols[ :, np.newaxis ] ** np.arange( 0, fitorder+1, 1 )
+        coltonminus1 = np.zeros( colton.shape )
+        coltonminus1[ :, 1: ] = cols[ :, np.newaxis ] ** np.arange( 0, fitorder, 1 )
+        coltonerr = np.zeros( colton.shape )
+        coltonerr[ :, 1: ] = np.arange( 1, fitorder+1, 1 ) * coltonminus1[ :, 1: ] * colerrs.value[ :, np.newaxis ]
+
+        zps = ( catdata['MAG_G'] - ( trns[ np.newaxis, : ] * colton ).sum( axis=1 )
+                + 2.5*np.log10( sourceflux ) )
+        zpvars = ( catdata[ 'MAGERR_G' ]**2 +
+                   ( trns[np.newaxis, : ] * coltonerr ).sum( axis=1 )**2 +
+                   ( 1.0857362 * sourcefluxerr / sourceflux )**2
+                  )
+
+        # Save these values so that tests outside can pull them and interrogate them
+        self.individual_zps = zps
+        self.individual_zpvars = zpvars
+        self.individual_cols = cols
+        self.individual_mags = catdata[ 'MAG_G' ]
+
+        wgood = ( ~np.isnan( zps ) ) & ( ~np.isnan( zpvars ) )
+        zps = zps[wgood]
+        zpvars = zpvars[wgood]
+        _logger.debug( f"{len(zps)} stars survived the not-NaN zeropoint check" )
+
+        zpval = np.sum( zps / zpvars ) / np.sum( 1. / zpvars )
+        dzpval = 1. / np.sum( 1. / (zpvars ) )
```

Comment:
If you have one really tiny error (by chance) it will dominate the value of `dzpval`. I think there must be a better way to calculate the uncertainty on `zpval`.

Reply:
Agreed. Although I need to think about this weighted sum thing some more.